### PR TITLE
Diff interpreter

### PR DESCRIPTION
**Pull request by supki**
#54 

Waiting for #67.

### CODE REVIEW

**Comment by supki** [comment]
@dmalikov, any opinion on this?

File: src/Control/Biegunka/Source/Git/Internal.hs [hunk 1]
```diff
@@ -9,24 +9,24 @@ module Control.Biegunka.Source.Git.Internal
   , Git (..)
   , actions, branch, failIfAhead
   , URI
-  , askGit, askGit_, updateGit
+  , runGit, updateGit
   , gitHash
   , defaultGit
   ) where
 
-import           Control.Monad                      (void, when)
```
Comment (dmalikov) [inline src/Control/Biegunka/Source/Git/Internal.hs]: Next time please prevent me from `stylish-haskell`ify imports if this is a format you want to keep them with

File: src/Control/Biegunka/Source/Git/Internal.hs [hunk 2]
```diff
@@ -91,41 +95,45 @@ git u p s = git' u p (actions s)
 git_ :: URI -> FilePath -> Script 'Sources ()
 git_ u p = git' u p mempty
 
-updateGit :: URI -> FilePath -> Git -> IO (Maybe String)
+updateGit :: URI -> FilePath -> Git -> IO (Maybe String, IO (Maybe String))
```
Comment (dmalikov) [inline src/Control/Biegunka/Source/Git/Internal.hs]: Why do we need this nested IO tuple?

File: src/Control/Biegunka/Source/Git/Internal.hs (hunk 1 above)
Comment (supki) [inline src/Control/Biegunka/Source/Git/Internal.hs]: Sure, no                                  problem                                                                       with that.

File: src/Control/Biegunka/Source/Git/Internal.hs (hunk 2 above)
Comment (supki) [inline src/Control/Biegunka/Source/Git/Internal.hs]: The outer IO action looks for the changes and gives them back, if any, as string together with the inner IO action that actually applies these changes. The reason for the second `Maybe String` is that the finalizer may have more information about the changes that actually happened (for instance, when cloning repositories)

**Comment by dmalikov** [comment]
It works pretty like it was expected in the original issue. Nice work!

Just a minor point,
```
haskell::[https://github.com/dmalikov/dotfiles]
  * file[/Users/m/.stylish-haskell.yaml] update (copy [/Users/m/git/dotfiles/configs/ghc/stylish-haskell.yaml])
    - contents changed from ‘d63fd9c7' to ‘d6fdaa8a’
      @@ -42,7 +42,7 @@
             # - compact: A more compact style.
             #
             # Default: vertical.
      -      style: vertical
      +      style: compact

             # stylish-haskell can detect redundancy of some language pragmas. If this
             # is set to true, it will remove those redundant pragmas. Default: true.
tmux::[https://github.com/dmalikov/dotfiles]
  * file[/Users/m/.tmux.conf] update (from template [/Users/m/git/dotfiles/configs/tmux/tmux.conf.template])
    - contents changed from ‘8c6c5234' to ‘2387d144’
vim/rc::[https://github.com/Shougo/neobundle.vim]
  * git source[/Users/m/.vim/bundle/neobundle.vim] update
    - ‘02bd438’ → ‘ed123ad’
git::[https://github.com/dmalikov/dotfiles]
  * git source[/Users/m/git/dotfiles] update
    - ‘cfa1ede’ → ‘6d0147b’
```

> - contents changed from ‘d63fd9c7' to ‘d6fdaa8a’
> - contents changed from ‘8c6c5234' to ‘2387d144’

Is there some better visualisation of the content update? These hashes looks almost like a git revisions which is kind of confusing. 
I don't think they are pretty informative neither for a copy action (when diff is actually displayed) nor for a substitute action (when all you have is this strange A -> B changelog). 
I guess the only thing that is interesting for a user is wether or not this file still exist, is it removed or is it  added.

**Comment by supki** [comment]
Visualization is shamelessly stolen from Chef, I'm afraid; I'm certainly not wedded to it.

Hash change with no diff looks strange. Was the line break at the end of the file removed? Generally, changes should be shown exactly the same way whether it's a substitution or a simple copy.

**Comment by dmalikov** [comment]
> Generally, changes should be shown exactly the same way whether it's a substitution or a simple copy.

Oh, nice. Yeah, probably it was an EOF. I just tried with proper change, diff was generated properly. :+1:

**Pull request closed without merging**